### PR TITLE
Add support for x87 registers on GISel register selection

### PR DESCRIPTION
**Pull request by MalaySanghi**
We handle 3 register classes for x87 - rfp32, rfp64 and rfp80.
1. X87 registers are assigned a pseudo register class. We need a new register bank for these classes.
2. We add bank information and enums for these.
3. Legalizer is updated to allow 32b and 64b even in absence of SSE. This is required because with SSE enabled, x86 doesn't use fp stack and instead uses SSE registers.
4. Functions in X86RegisterBankInfo need to decide whether to use the pseudo classes or SSE-enabled classes. I add MachineInstr as an argument to static helper function getPartialMappingIdx to this end.

Add/Update tests.

### CODE REVIEW

**Comment by MalaySanghiIntel** [comment]
Tagging @e-kud for review. Please add others for opinion/review.

File: llvm/test/CodeGen/X86/GlobalISel/regbankselect-x87.ll [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+  ; X64-NEXT:   [[COPY2:%[0-9]+]]:psr(s32) = COPY [[LOAD]](s32)
+  ; X64-NEXT:   [[FADD:%[0-9]+]]:psr(s32) = G_FADD [[COPY2]], [[C]]
+  ; X64-NEXT:   [[COPY3:%[0-9]+]]:gpr(s32) = COPY [[FADD]](s32)
+  ; X64-NEXT:   G_STORE [[COPY3]](s32), [[COPY1]](p0) :: (store (s32) into %ir.1)
+  ; X64-NEXT:   RET 0
+  %3 = load float, ptr %0
+  %4 = fadd float %3, 20.0
+  store float %4, ptr %1
+  ret void
+}
```
[the remark this answers is not in the document]
Comment (MalaySanghiIntel) [reply llvm/test/CodeGen/X86/GlobalISel/regbankselect-x87.ll:237]: No. We (me and @e-kud) plan to have more instructions in a separate patch.

**Comment by tschuett** [comment]
```
This is required because with SSE enabled, x86 doesn't use fp stack and instead uses AVX registers.
```
Did you mean SSE registers?

**Comment by MalaySanghiIntel** [comment]
> ```
> This is required because with SSE enabled, x86 doesn't use fp stack and instead uses AVX registers.
> ```
> 
> Did you mean SSE registers?

Yes, I did. Updated. Thanks for pointing this out

File: llvm/lib/Target/X86/GISel/X86RegisterBankInfo.cpp [hunk 2]
```diff
@@ -234,7 +250,7 @@ X86RegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
     (void)Size;
     assert((Size == 32 || Size == 64) && "Unsupported size for G_FCMP");
 
-    auto FpRegBank = getPartialMappingIdx(Ty1, /* isFP */ true);
+    auto FpRegBank = getPartialMappingIdx(MI, Ty1, /* isFP */ true);
```
[the remark this answers is not in the document]
Comment (MalaySanghiIntel) [reply llvm/lib/Target/X86/GISel/X86RegisterBankInfo.cpp:253]: Are you asking me to add a "=" after isFP or was this a mistake?

**Comment by MalaySanghiIntel** [comment]
FYI, the windows failure is random. Related issue: https://github.com/llvm/llvm-project/issues/83534

File: llvm/test/CodeGen/X86/GlobalISel/regbankselect-x87.ll [hunk 3]
```diff
@@ -0,0 +1,237 @@
+; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
+; RUN: llc < %s -mtriple=i686-- -mattr=+x87,-sse,-sse2 -global-isel -stop-after=regbankselect | FileCheck %s --check-prefix=X32
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/GlobalISel/regbankselect-x87.ll:2]: (pedantic) Use X86 - X32 should be used only for gnux32 targets

**Review by RKSimon — COMMENTED** [review]
Does GISel have support for the stackifier?

**Comment by MalaySanghiIntel** [comment]
> Does GISel have support for the stackifier?

The stackifier is an independent post RA pass. It is enabled on the GISel path as weel. However, it doesn't work as expected today and many x87 programs fail at instruction selection in GISel.

I'm not sure what you mean by support here - stackifier is pattern based i.e. it requires the opcodes from `OpcodeTable` in the IR. On the GISel path, these instructions are not yet generated (and it is WIP). The current PR is the first step. I plan to have follow up changes which will improve the porting of patterns from DAG to GISel. That will enable us to get instructions like `LD_Fp32m` and we'll be able to evaluate stackifier status better.

**Comment by MalaySanghiIntel** [comment]
> Does GISel have support for the stackifier?

When you say stackifier, do you mean the post-RA pass or something else? I don't recall anything seeing anything specific to stackifier in the DAG passes. Maybe I missed something. Can you please explain if I missed it?

**Comment by RKSimon** [comment]
> Does GISel have support for the stackifier?

Sorry that was a vague question - but I think you answered it :)

File: llvm/lib/Target/X86/GISel/X86RegisterBankInfo.cpp [hunk 4]
```diff
@@ -219,8 +235,8 @@ X86RegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
 
     bool FirstArgIsFP = Opc == TargetOpcode::G_SITOFP;
     bool SecondArgIsFP = Opc == TargetOpcode::G_FPTOSI;
-    OpRegBankIdx[0] = getPartialMappingIdx(Ty0, /* isFP */ FirstArgIsFP);
-    OpRegBankIdx[1] = getPartialMappingIdx(Ty1, /* isFP */ SecondArgIsFP);
+    OpRegBankIdx[0] = getPartialMappingIdx(MI, Ty0, /* isFP= */ FirstArgIsFP);
```
Comment (arsenm) [inline llvm/lib/Target/X86/GISel/X86RegisterBankInfo.cpp:238]: I think these need to be hit with clang-format again

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Target/X86/GISel/X86RegisterBankInfo.cpp (hunk 4 above)
Comment (MalaySanghiIntel) [reply llvm/lib/Target/X86/GISel/X86RegisterBankInfo.cpp:238]: These are 78 and 79 char long, clang-format didn't split these.

**Comment by e-kud** [comment]
LGTM. @MalaySanghiIntel thanks!

**Review by e-kud — APPROVED** [review]

**Comment by e-kud** [comment]
Ashes on my head, I haven't fixed the tags. I will be more attentive next time.

**Comment by MalaySanghiIntel** [comment]
> This commit is associated with a `users.noreply.github.com` email address. Per https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it/74223/32 , proper email addresses are preferred.

I've made my email public. Thanks for pointing this out.

**Pull request merged**